File: backend-nestjs/src/modules/damage-reports/application/queries/get-damage-report-by-id/get-damage-report-by-id.handler.ts
import { IQueryHandler, QueryHandler } from '@nestjs/cqrs';
import { NotFoundException } from '@nestjs/common';
import { GetDamageReportByIdQuery } from './get-damage-report-by-id.query';
import { DamageReportRepository } from '../../../infrastructure/damage-report.repository';
import { DamageReport } from '../../../domain/damage-report';

@QueryHandler(GetDamageReportByIdQuery)
export class GetDamageReportByIdHandler implements IQueryHandler<
  GetDamageReportByIdQuery,
  DamageReport
> {
  constructor(private readonly repository: DamageReportRepository) {}

  async execute(query: GetDamageReportByIdQuery): Promise<DamageReport> {
    const report = await this.repository.findById(query.id);
    if (!report) {
      throw new NotFoundException(
        `Damage report with ID '${query.id}' not found`,
      );
    }
    return report;
  }
}

File: backend-springboot/src/main/java/com/greenalgeria/damagereport/application/query/GetDamageReportByIdHandler.java
package com.greenalgeria.damagereport.application.query;

import com.greenalgeria.damagereport.application.*;
import com.greenalgeria.damagereport.domain.DamageReportRepository;
import com.greenalgeria.shared.cqrs.QueryHandler;
import java.util.Optional;
import org.springframework.stereotype.Component;

@Component
public class GetDamageReportByIdHandler
        implements QueryHandler<GetDamageReportByIdQuery, Optional<DamageReportResponse>> {

    private final DamageReportRepository damageReportRepository;

    public GetDamageReportByIdHandler(DamageReportRepository damageReportRepository) {
        this.damageReportRepository = damageReportRepository;
    }

    public Optional<DamageReportResponse> handle(GetDamageReportByIdQuery query) {
        return damageReportRepository.findById(query.id()).map(DamageReportResponse::from);
    }

    @Override
    public Class<GetDamageReportByIdQuery> supportedQuery() {
        return GetDamageReportByIdQuery.class;
    }
}
